setup.py

```diff
@@ -956,11 +1024,91 @@ def build_torch_extension_v2(build_ext, options, torch_version):
+
+    # check again if the cmake binary is usable
+    try:
+        subprocess.check_output([cmake_bin, '--version'])
+    except OSError as e:
+        raise RuntimeError('Check CMake executable failed: {}'.format(str(e)))
+
+    # Statically linked archive files go into the provided output directory
+    extdir = os.path.abspath(
+        os.path.dirname(build_ext.get_ext_fullpath(ext.name)))
+    config = 'Debug' if build_ext.debug else 'Release'
+    cmake_args = [
+        '-DUSE_MPI=ON',
+        '-DCMAKE_BUILD_TYPE=' + config,
+        '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY_{}={}'.format(config.upper(), extdir),
+        '-DCMAKE_ARCHIVE_OUTPUT_DIRECTORY_{}={}'.format(config.upper(),
+                                                        output_dir),
+    ]
+    cmake_build_args = [
+        '--config', config,
+        '--', '-j4',
+    ]
+
+    # Keep temp build files within a unique subdirectory
+    build_temp = os.path.abspath(os.path.join(build_ext.build_temp, ext.name))
+    if not os.path.exists(build_temp):
+        os.makedirs(build_temp)
+
+    # Config and build the extension
+    try:
+        subprocess.check_call([cmake_bin, ext.cmake_lists_dir] + cmake_args,
+                              cwd=build_temp)
+        subprocess.check_call([cmake_bin, '--build', '.'] + cmake_build_args,
+                              cwd=build_temp)
+    except OSError as e:
+        raise RuntimeError('CMake failed: {}'.format(str(e)))
+    # Add the library so other extensions will link against it during compilation
+    options['LIBRARIES'] += [ext.name]
+
+
 # run the customize_compiler
 class custom_build_ext(build_ext):
     def build_extensions(self):
         options = get_common_options(self)
         built_plugins = []
+
+        # All statically linked libraries will be placed here
+        lib_output_dir = os.path.abspath(os.path.join(self.build_temp, 'lib'))
+        if not os.path.exists(lib_output_dir):
+            os.makedirs(lib_output_dir)
+        options['LIBRARY_DIRS'] += [lib_output_dir]
+
+        # Build CMake libraries first so they can be statically linked against extensions
+        for ext in self.extensions:
+            if isinstance(ext, CMakeExtension):
+                # TODO: install cmake in local environment after entry point issue
+                #  has some updates.
+                #  https://github.com/scikit-build/cmake-python-distributions/issues/80
+                if ext.name == 'gloo':
+                    if is_mac:
+                        print(
+                            'INFO: Submodule Gloo cannot compile on MacOS, skip compiling '
+                            'Gloo.')
+                    elif not have_cmake:
+                        print(
+                            'INFO: Submodule Gloo cannot compile without cmake, '
+                            'skip compiling Gloo.')
+                    else:
+                        build_cmake(self, ext, lib_output_dir, options)
```

Comment:
Should we just build cmake directly w/o iterating extensions?

Reply:
Yes. Fixed.